hadoop-hdds/docs/content/design/s3-performance.md

```diff
@@ -0,0 +1,205 @@
+
+```
+public class UserGroupInformation {
+  
+  ...
+  
+  @Override
+  public boolean equals(Object o) {
+    if (o == this) {
+      return true;
+    } else if (o == null || getClass() != o.getClass()) {
+      return false;
+    } else {
+      return subject == ((UserGroupInformation) o).subject;
+    }
+  }
+}
+```
+
+ But the UGI initialization of Ozone always creates a new `Subject` instance for each request (even if the subject name is the same). In `OzoneClientProducer`:
+
+```
+  UserGroupInformation remoteUser =
+          UserGroupInformation.createRemoteUser(awsAccessId); // <-- new Subject is created
+      
+      if (OzoneSecurityUtil.isSecurityEnabled(config)) {
+        try {
+          OzoneTokenIdentifier identifier = new OzoneTokenIdentifier();
+          //setup identifier
+          
+          Token<OzoneTokenIdentifier> token = new Token(identifier.getBytes(),
+              identifier.getSignature().getBytes(UTF_8),
+              identifier.getKind(),
+              omService);
+          remoteUser.addToken(token);
+          ....
+```
+
+**As a result Hadoop RPC caching doesn't apply to Ozone S3g**. It's a good news because it's secure, but bad news as the performance is bad.
+
+# Proposed change
+
+We need an RPC mechanism between the Ozone S3g service and Ozone Manager service which can support per-request authentication and accepts
+
+The  Ozone Manager client already has a pluggable transport interface: `OmTransport` is a simple interface which can deliver `OMRequest` messages:
+
+```
+public interface OmTransport {
+
+  /**
+   * The main method to send out the request on the defined transport.
+   */
+  OMResponse submitRequest(OMRequest payload) throws IOException;
+  ...
+```
+
+ The proposal is to create a new **additional** transport, based on GRPC, which can do the per-request authentication. **Existing Hadoop clients will use the well-known Hadoop RPC client**, but S3g can start to use this specific transport to achieve better performance.
+
+As this is nothing more, just a transport: exactly the same messages (`OmRequest`) will be used, it's not a new RPC interface.
+
+Only one modification is required in the RPC interface: a new per-request`token` field should be introduced in `OMRequest` which is optional.
+
+A new GRPC service should be started in Ozone Manager, which receives `OMRequest` and for each request, the Hadoop `UserGroupInformation` is set based on the new token field (after authentication).
+
+`OzoneToken` identifier can be simplified (after deprecation period) with removing the S3 specific part, as it won't be required any more.
+
+With this approach the `OzoneClient` instances can be cached on S3g side (with persistent GRPC connections) as the authentication information is not part of the OzoneClient any more (added by the `OmTransport` implementation per request (in case of GRPC) or per connection (in case of HadoopRPC)).
```

Comment:
Does that mean with this approach we need one ozone Client instantiated, as token is part of OMRequest.

Reply:
Few questions:
This means S3G does not use hadoop Rpc Client, it will be use GrpcClient
So how OM HA will be handled retry handling logic, so all that logic need to be implemented in this new GrpcClient?
And once the token is validated, will it go with the normal flow of execution in OzoneManager?

Few minor questions, as I don't have much expertise on Grpc Implementation.
1. Does GrpcServer also will have RPC handler threads where requests can be handled parallel on OM.

Reply:
> Does that mean with this approach we need one ozone Client instantiated, as token is part of OMRequest.

Yes. 

>  And once the token is validated, will it go with the normal flow of execution in OzoneManager?

Yes, exactly the same logic.

> Does GrpcServer also will have RPC handler threads where requests can be handled parallel on OM.

Yes. As far as I understood from the documentation the new thread is created by the async IO handler thread. 

But if we need more freedom, we can always introduce a simple Executor.

But it's a very good question. Thinking about this, I have new ideas: with separating S3g and client side traffic we can monitor the two in different way (for example compare queue time of client and s3g calls, or set priorities). Not in this step, but something which will be possible.

> How OM HA will be handled retry handling logic, so all that logic need to be implemented in this new GrpcClient?

Yes. We need to take care about the retry logic. My initial suggestion is to create 3, persistent connection to all the OM, and in case of not leader exception try to send the message on a different connection.

In case of client it can be expensive (always create 3 different connection to 3 different OM HA instance), but in case of S3g and persistent connections it seems to more effective as the connections are persistent.

Reply:
So, it is like a new retry logic should be implemented for GrpcClient.

Reply:
Thank You for detailed answers for other points.

>I have new ideas: with separating S3g and client side traffic we can monitor the two in different way (for example >compare queue time of client and s3g calls, or set priorities). Not in this step, but something which will be possible.

This idea looks interesting, but at end, both are coming from end clients, so getting additional metrics helps to understand better calls from S3/other interface, but not sure in which scenarios this will help.

Reply:
For example to understand / compare the cluster usage. Which part is used more HCFS or s3? What is the source of small files s3 or HCFS? 

This (different metrics) is not something to do right now but an interesting option to think forward if this approach is accepted.

Reply:
> So, it is like a new retry logic should be implemented for GrpcClient.

Yes. And I argue that this logic can be optimized for servers (connections to different OM instances can be cached long-term) and not only optimized for client (open second connections to the right OM only in case of leader election)